dm/worker/upgrade.go

```diff
@@ -0,0 +1,228 @@
+		return errors.Trace(ErrInValidHandler)
+	}
+
+	data, err := ver.MarshalBinary()
+	if err != nil {
+		return errors.Annotatef(err, "marshal version %v to binary data", ver)
+	}
+
+	err = h.Put(dmWorkerVersionKey, data, nil)
+	return errors.Annotatef(err, "save version %v into levelDB with key %s", ver, dmWorkerVersionKey)
+}
+
+// tryUpgrade tries to upgrade from an older version.
+func tryUpgrade(dbDir string) error {
+	// 1. check the DB directory
+	notExist := false
+	fs, err := os.Stat(dbDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			notExist = true
+		} else {
+			return errors.Annotatef(err, "get stat for %s", dbDir)
+		}
+	} else if !fs.IsDir() { // should be a directory
+		return errors.NotValidf("directory %s for DB", dbDir)
+	}
+
+	// 2. open the kv DB
+	db, err := openDB(dbDir, defaultKVConfig)
+	if err != nil {
+		return errors.Annotatef(err, "open DB for %s", dbDir)
+	}
+	defer func() {
+		err = db.Close()
+		if err != nil {
+			log.Errorf("[worker upgrade] close DB fail %v", err)
+		}
+	}()
+
+	if notExist {
+		log.Infof("[worker upgrade] no previous operation log exists, no need to upgrade")
+		// still need to save the current version version
+		currVer := newCurrentVersion()
+		err = saveVersion(db, currVer)
+		return errors.Annotatef(err, "save current version %v into DB %s", currVer, dbDir)
+	}
+
+	// 3. load previous version
+	prevVer, err := loadVersion(db)
+	if err != nil {
+		return errors.Annotatef(err, "load previous version from DB %s", dbDir)
+	}
+	log.Infof("[worker upgrade] the previous version is %v", prevVer)
+
+	// 4. check needing to upgrade
+	currVer := newCurrentVersion()
+	if prevVer.compare(currVer) == 0 {
+		log.Infof("[worker upgrade] the previous and current versions both are %v, no need to upgrade", prevVer)
+		return nil
+	} else if prevVer.compare(currVer) > 0 {
+		log.Warnf("[worker upgrade] the previous version %v is newer than current %v, automatic downgrade is not supported now", prevVer, currVer)
+		return nil
+	}
+
+	// 5. upgrade from previous version to +1, +2, ...
+	if prevVer.compare(workerVersion1) < 0 {
+		err = upgradeToVer1(db)
+		if err != nil {
+			return errors.Annotatef(err, "upgrade to version %v", workerVersion1)
+		}
+	}
+
+	// 6. save current version after upgrade done
+	err = saveVersion(db, currVer)
+	return errors.Annotatef(err, "save current version %v into DB %s", currVer, dbDir)
+}
+
+// upgradeToVer1 upgrades from version 0 to version 1.
+// before this version, we use `LittleEndian` to encode/decode operation log ID, but it's not correct when scanning operation log by log ID.
+// so, if upgrading from previous version to this one, we need to:
+//  1. remove all operation log in the levelDB
+//  2. reset handled pointer
+//  3. remove all task meta in the levelDB
+// and let user to restart all necessary tasks.
+func upgradeToVer1(db *leveldb.DB) error {
+	log.Infof("[worker upgrade] upgrading to version %v", workerVersion1)
+	txn, err := db.OpenTransaction()
```

Comment:
don't use transaction here, use db directly

Reply:
changed in https://github.com/pingcap/dm/pull/187/commits/10281b801665c53de0201cd2cb9a1a82dc2ccec7.